sklearn/tree/tests/test_monotonic_tree.py

```diff
@@ -0,0 +1,348 @@
+    # should have:
+    #
+    #       root
+    #      /    \
+    #     a      b
+    #    / \    / \
+    #   c   d  e   f
+    #
+    # c <= d <= root <= e <= f
+
+    rng = np.random.RandomState(global_random_seed)
+    n_samples = 1000
+    n_features = 1
+    X = rng.rand(n_samples, n_features)
+    y = rng.rand(n_samples)
+
+    if depth_first:
+        # No max_leaf_nodes, default depth first tree builder
+        clf = DecisionTreeRegressor(
+            splitter=splitter,
+            monotonic_cst=[monotonic_sign],
+            random_state=global_random_seed,
+        )
+    else:
+        # max_leaf_nodes triggers best first tree builder
+        clf = DecisionTreeRegressor(
+            splitter=splitter,
+            monotonic_cst=[monotonic_sign],
+            max_leaf_nodes=n_samples,
+            random_state=global_random_seed,
+        )
+    clf.fit(X, y)
+
+    assert_1d_reg_tree_children_monotonic_bounded(clf.tree_, monotonic_sign)
+    assert_1d_reg_monotonic(clf, monotonic_sign, np.min(X), np.max(X), 100)
+
+
+def assert_nd_reg_tree_children_monotonic_bounded(tree_, monotonic_cst):
+    upper_bound = np.full(tree_.node_count, np.inf)
+    lower_bound = np.full(tree_.node_count, -np.inf)
+    for i in range(tree_.node_count):
+        feature = tree_.feature[i]
+        assert tree_.value[i] <= upper_bound[i]
+        assert tree_.value[i] >= lower_bound[i]
+        if feature < 0:
+            # Leaf: nothing to do
+            continue
+        else:
+            i_left = tree_.children_left[i]
+            i_right = tree_.children_right[i]
+            if monotonic_cst[feature] == 0:
+                # Feature without monotonicity constraint: propagate bounds
+                # down the tree to both children.
+                # Otherwise with 2 features and a POS constraint on feature 0
+                # the following tree can be accepted, although it does not
+                # respect the positive monotonicity constraint:
+                #
+                #                      X[0] <= 0
+                #                      value = 100
+                #                     /            \
+                #          X[0] <= -1                X[1] <= 0
+                #          value = 50                value = 150
+                #        /            \             /            \
+                #    leaf           leaf           leaf          leaf
+                #    value = 25     value = 75     value = 50    value = 250
+
+                upper_bound[i_left] = upper_bound[i]
+                lower_bound[i_left] = lower_bound[i]
+                upper_bound[i_right] = upper_bound[i]
+                lower_bound[i_right] = lower_bound[i]
+            else:
+                # Feature with constraint: check monotonicity
+                assert (
+                    monotonic_cst[feature] * tree_.value[i_left]
+                    <= monotonic_cst[feature] * tree_.value[i_right]
```

Comment:
Again, I think the logic with multiplying by the monotonic_cst to avoid case distinction does not work as intended.

Reply:
This works as far as equality is involved for the comparison, but I think that this is not super explicit.